### PR TITLE
FEDX-142: Migrate to null safety

### PR DESCRIPTION
**Pull request by alishadalal-wk**
Migrate to null safety. Run the null safety migration tool with Dart migrate and check/adjust its suggestions if needed.
Apply the migration, and fix any static errors. Do a dart format, run the tests, fixing any issues.

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/dart/lib/src/frugal/f_context.dart [hunk 1]
```diff
@@ -49,8 +49,8 @@ final Duration _defaultTimeout = new Duration(seconds: 5);
 class FContext {
   static int _globalOpId = 0;
 
-  Map<String, String> _requestHeaders;
-  Map<String, String> _responseHeaders;
+  late Map<String, String?> _requestHeaders;
+  late Map<String, String?> _responseHeaders;
```
Comment (brettkail-wk) [inline lib/dart/lib/src/frugal/f_context.dart:53]: I believe Headers.encode/decodeFromFrame require the values to be non-null, so can these be made non-null?

As a style question, is it better to assign these to `{}` here rather than `late`?

File: lib/dart/lib/src/frugal/f_middleware.dart [hunk 2]
```diff
@@ -29,9 +29,9 @@ typedef InvocationHandler Middleware(InvocationHandler handler);
 /// Contains an [InvocationHandler] used to proxy the given service method
 /// This should only be used by generated code.
 class FMethod {
-  String _serviceName;
-  String _methodName;
-  InvocationHandler _handler;
+  String? _serviceName;
+  String? _methodName;
```
Comment (brettkail-wk) [inline lib/dart/lib/src/frugal/f_middleware.dart:33]: Can these be made non-null?  Can `_handler` be initialized so it doesn't need to be `late`?  Pre-existing, can all the members be made `final`?

Then, can `InvocationHandler` be changed to use non-null parameters?  (I'm less familiar with Dart return type semantics, but `Future?` seems like a surprising return value.)

File: lib/dart/lib/src/frugal/f_middleware.dart [hunk 3]
```diff
@@ -66,8 +66,8 @@ class FMethod {
 
 /// [Middleware] for debugging that logs the requests and responses in json
 /// format.
-InvocationHandler debugMiddleware(InvocationHandler next) {
-  return (String serviceName, String methodName, List<Object> args) async {
+InvocationHandler debugMiddleware(InvocationHandler? next) {
```
Comment (brettkail-wk) [inline lib/dart/lib/src/frugal/f_middleware.dart:69]: Can this be made non-null?  It implicitly was before.

I've stopped going through the PR line-by-line looking for non-nullable opportunities.  It seems like it would be valuable to make more things non-nullable where null would already cause an exception?

File: lib/dart/lib/src/frugal/transport/f_http_transport.dart [hunk 4]
```diff
@@ -123,20 +123,22 @@ class FHttpTransport extends FTransport {
...
       throw new TTransportError(FrugalTTransportErrorType.UNKNOWN, ex.message);
     }
 
     // Attempt to decode the response payload
-    Uint8List data;
+    late Uint8List data;
     try {
-      data = new Uint8List.fromList(base64.decode(response.body.asString()));
+      var strBody = response.body.asString();
+      if (strBody != null)
```
Comment (brettkail-wk) [inline lib/dart/lib/src/frugal/transport/f_http_transport.dart:140]: It seems like `asString()` should be non-nullable?  Even if it is, I think `.decode` currently fails if the param is null, so it seems like this should just be `.asString()!` and simplified?

**Comment by robbecker-wf** [comment]
Let's get the dart format --fix changes merged to develop first, that should clean up this PR. https://github.com/Workiva/frugal/pull/2279

File: lib/dart/lib/src/frugal/f_context.dart [hunk 5]
```diff
@@ -85,7 +85,8 @@ class FContext {
   /// The request timeout for any method call using this context.
   /// The default is 5 seconds.
   Duration get timeout {
-    return Duration(milliseconds: int.parse(_requestHeaders[_timeoutHeader]));
+    return Duration(
+        milliseconds: int.parse(_requestHeaders[_timeoutHeader] ?? ''));
```
[the remark this answers is not in the document]
Comment (brettkail-wk) [reply lib/dart/lib/src/frugal/f_context.dart:89]: The rest of the code ensures the key always exists in the map, so it seems like `int.parse(_requestHeaders[_timeoutHeader]!)` would be simplest and consistent with what the code already does.

File: lib/dart/lib/src/frugal/f_context.dart [hunk 6]
```diff
@@ -94,11 +95,11 @@ class FContext {
   }
 
   /// Correlation id for the context.
-  String get correlationId => _requestHeaders[_cidHeader];
+  String? get correlationId => _requestHeaders[_cidHeader];
 
   /// The operation id for the context.
   int get _opId {
-    var opIdStr = _requestHeaders[_opidHeader];
+    var opIdStr = _requestHeaders[_opidHeader] ?? '';
```
[the remark this answers is not in the document]
Comment (brettkail-wk) [reply lib/dart/lib/src/frugal/f_context.dart:102]: Similarly here, it seems like `_requestHeaders[_opidHeader]!` would be consistent with the current logic.

File: lib/dart/lib/src/frugal/f_context.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/f_context.dart]: Updated it to `_requestHeaders[_timeoutHeader]!`

File: lib/dart/lib/src/frugal/f_context.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/f_context.dart]: updated it to `_requestHeaders[_opidHeader]!`

File: lib/dart/lib/src/frugal/f_context.dart [hunk 7]
```diff
@@ -49,8 +49,8 @@ final Duration _defaultTimeout = Duration(seconds: 5);
 class FContext {
   static int _globalOpId = 0;
 
-  Map<String, String> _requestHeaders;
-  Map<String, String> _responseHeaders;
+  Map<String, String?> _requestHeaders = {};
+  Map<String, String?> _responseHeaders = {};
```
[the remark this answers is not in the document]
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/f_context.dart]: Updated from `Map<String, String?>` to `Map<String, String>`

File: lib/dart/lib/src/frugal/f_context.dart [hunk 8]
```diff
@@ -120,17 +121,17 @@ class FContext {
...
   /// Get the named request header.
-  String requestHeader(String name) {
+  String? requestHeader(String name) {
     return _requestHeaders[name];
   }
 
   /// Get requests headers map.
-  Map<String, String> requestHeaders() {
+  Map<String, String?> requestHeaders() {
     return UnmodifiableMapView(_requestHeaders);
```
[the remark this answers is not in the document]
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/f_context.dart]: Changed `Map<String, String?> requestHeaders()` to `Map<String, String> requestHeaders()`

File: lib/dart/lib/src/frugal/transport/f_async_transport.dart [hunk 9]
```diff
@@ -37,7 +37,12 @@ abstract class FAsyncTransport extends FTransport {
   }
 
   @override
-  Future<TTransport> request(FContext ctx, Uint8List payload) async {
+  Future<TTransport> request(FContext? ctx, Uint8List? payload) async {
```
[the remark this answers is not in the document]
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/transport/f_async_transport.dart:40]: It is because in the file **f_adapter_transport_test.dart** on line no. 85 the code `await transport.request(null, null); `passed null params. So the test gets failed if we don't make it nullable.

```
Error: The value 'null' can't be assigned to the parameter type 'FContext' because 'FContext' is not nullable.
   - 'FContext' is from 'package:frugal/src/frugal.dart' ('lib/src/frugal.dart').
          await transport.request(null, null);
                                  ^
Error: The value 'null' can't be assigned to the parameter type 'Uint8List' because 'Uint8List' is not nullable.
   - 'Uint8List' is from 'dart:typed_data'.
          await transport.request(null, null);

```

File: lib/dart/lib/src/frugal/transport/base_f_transport_monitor.dart [hunk 10]
```diff
@@ -66,20 +66,21 @@ class BaseFTransportMonitor extends FTransportMonitor {
   }
 
   @override
-  int onClosedUncleanly(Object cause) {
+  int? onClosedUncleanly(Object? cause) {
```
[the remark this answers is not in the document]
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/transport/base_f_transport_monitor.dart:69]: We need to add `?` because some test files contain `when(monitor.onClosedUncleanly(any)).thenReturn(1);` 
So, 
when Object cause is non-nullable, **tests get failed** due to the reason `The argument type 'Null' can't be assigned to the parameter type 'Object' because 'Object' is not nullable.`

Also, added ? for in the return type(int?) because **tests get failed** due to reason - `type 'Null' is not a subtype of type 'int'`

Same reason for **onReopenFailed** as tests contains - `when(monitor.onReopenFailed(any, any)).thenReturn(1);` which fails if we don't add `?`

File: lib/dart/lib/src/frugal/transport/f_async_transport.dart [hunk 11]
```diff
@@ -70,10 +75,10 @@ abstract class FAsyncTransport extends FTransport {
       // don't wait until this is disposed to cancel these
       await closedSub.cancel();
       if (!closedCompleter.isCompleted) {
-        closedCompleter.complete();
+        closedCompleter.complete(Uint8List(0));
```
[the remark this answers is not in the document]
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/transport/f_async_transport.dart:78]: I am not sure about this, but the **complete** calls in the **finally** block of the code ensure that certain **Completer** objects are resolved properly, regardless of whether the main code block completes normally or due to an exception. 

@robbecker-wf , do you have any idea about this?

File: lib/dart/lib/src/frugal/f_context.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/f_context.dart]: Ok. I am thinking for following solution for optimization. Please have a look.

**Define a variable:**

`UnmodifiableMapView<String, String>? _requestHeadersCache;
`

**When adding a request header, invalidate the cache:**
```
  void addRequestHeader(String name, String value) {
    _requestHeaders[name] = value;
    _requestHeadersCache = null;  // Invalidate the cache when adding a new
  }
```
  
 **When getting request headers, add to cache if its null and return data from it:**
 
```
  Map<String, String> requestHeaders() {
    if (_requestHeadersCache == null) {
      _requestHeadersCache = UnmodifiableMapView(_requestHeaders);
    }
    return _requestHeadersCache!;
  }
```
  
Also same can be applied for caching the response headers as well.   @robbecker-wf , Is it ok and Should we apply it OR any other suggestions?

File: lib/dart/lib/src/frugal/transport/base_f_transport_monitor.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (brettkail-wk) [reply lib/dart/lib/src/frugal/transport/base_f_transport_monitor.dart:69]: Sorry, I don't have enough Dart experience, but there must be some way to make the tests work without affecting the signatures of non-test code.

Searching Google, I find [mocktail](https://pub.dev/packages/mocktail), which has a `when` that wraps the invocation so that the null result works.  Perhaps @robbecker-wf has an idea of what has been done to solve this in other repos.

File: lib/dart/lib/src/frugal/transport/f_async_transport.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (brettkail-wk) [reply lib/dart/lib/src/frugal/transport/f_async_transport.dart:40]: It seems the tests should be fixed.

File: lib/dart/lib/src/frugal/f_context.dart [hunk 12]
```diff
@@ -49,8 +49,8 @@ final Duration _defaultTimeout = Duration(seconds: 5);
 class FContext {
   static int _globalOpId = 0;
 
-  Map<String, String> _requestHeaders;
-  Map<String, String> _responseHeaders;
+  Map<String, String> _requestHeaders = {};
+  Map<String, String> _responseHeaders = {};
```
Comment (brettkail-wk) [inline lib/dart/lib/src/frugal/f_context.dart:53]: This is fine, but pre-existing, it would be nice to make these `final` (and modify the constructors to copy into the map).

File: lib/dart/lib/src/frugal/f_context.dart [hunk 13]
```diff
@@ -120,12 +120,12 @@ class FContext {
       return;
     }
     for (var name in headers.keys) {
-      _requestHeaders[name] = headers[name];
+      _requestHeaders[name] = headers[name]!;
```
Comment (brettkail-wk) [inline lib/dart/lib/src/frugal/f_context.dart:123]: This is fine, but pre-existing, but `_requestHeaders.addAll(headers)` would remove the need for `!`, and similarly in `addResponseHeaders`.

Re: `Headers.encode`, I don't know the generate Dart pattern for iterating a `Map` in a null-safe way.  The `!` there is somewhat unsatisfying.  Iterating `headers.entries` or `forEach` would fix that, but I'm not sure about the performance trade-offs (redundant `[]` lookup vs `MapEntry` object creation vs closure).

File: lib/dart/lib/src/frugal/transport/monitor_runner.dart [hunk 14]
```diff
@@ -75,22 +75,22 @@ class MonitorRunner extends Disposable {
...
+    _wait = _monitor.onClosedUncleanly(cause) ?? -1;
     if (_wait < 0) {
       _log.log(Level.WARNING, 'instructed not to reopen');
       _stop(failed: true);
       return;
     }
     _reopenCompleter = Completer();
     _startReopenTimer();
-    await _reopenCompleter.future;
+    await _reopenCompleter?.future;
```
Comment (brettkail-wk) [inline lib/dart/lib/src/frugal/transport/monitor_runner.dart:93]: This looks conceptually wrong.  It seems like either `final reopenCompleter = Completer(); _reopenCompleter = reopenCompleter` or use `!` here?

File: lib/dart/lib/src/frugal/f_context.dart (hunk 12 above)
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/f_context.dart]: Ok. I have made these final and updated the code related to it as needed. Please have a look.

```
  final Map<String, String> _requestHeaders;
  final Map<String, String> _responseHeaders;

  /// Create a new [FContext] with the optionally specified [correlationId].
  FContext({String correlationId = ""})
      : _requestHeaders = {},
        _responseHeaders = {} {
    if (correlationId == "") {
      correlationId = _generateCorrelationId();
    }
    _globalOpId++;
    _requestHeaders[_cidHeader] = correlationId;
    _requestHeaders[_opidHeader] = _globalOpId.toString();
    _requestHeaders[_timeoutHeader] = _defaultTimeout.inMilliseconds.toString();
  }

  /// Create a new [FContext] with the given request headers.
  FContext.withRequestHeaders(Map<String, String> headers)
      : _requestHeaders = Map.from(headers),
        _responseHeaders = {} {
    if (!_requestHeaders.containsKey(_cidHeader) ||
        _requestHeaders[_cidHeader] == "") {
      _requestHeaders[_cidHeader] = _generateCorrelationId();
    }
    if (!_requestHeaders.containsKey(_opidHeader) ||
        _requestHeaders[_opidHeader] == "") {
      _globalOpId++;
      _requestHeaders[_opidHeader] = _globalOpId.toString();
    }
    if (!_requestHeaders.containsKey(_timeoutHeader) ||
        _requestHeaders[_timeoutHeader] == "") {
      _requestHeaders[_timeoutHeader] =
          _defaultTimeout.inMilliseconds.toString();
    }
  }
```

File: lib/dart/lib/src/frugal/f_context.dart (hunk 13 above)
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/f_context.dart:126]: Yes, `_requestHeaders.addAll(headers)` would remove the need for `!` 

Using headers.entries, We can update code as following:

```
for (var entry in headers.entries) {
  _requestHeaders[entry.key] = entry.value;
}
```

I think, it might be slightly more performant as it doesn't need the extra map lookup for the each key.

If we need to add all the entries we can go with first approach and if we don't then with the second one. 
@robbecker-wf, can you please suggest?

File: lib/dart/lib/src/frugal/transport/monitor_runner.dart (hunk 14 above)
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/transport/monitor_runner.dart]: Updated it to `await _reopenCompleter!.future;`

File: lib/dart/lib/src/frugal/transport/f_transport.dart [hunk 15]
```diff
@@ -75,8 +75,8 @@ abstract class FTransport extends Disposable {
     }
 
     if (requestSizeLimit != null &&
-        requestSizeLimit > 0 &&
-        payload.length > requestSizeLimit) {
+        requestSizeLimit! > 0 &&
```
[the remark this answers is not in the document]
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/transport/f_transport.dart:78]: I didn't quite understand. Could you please explain it a bit more?

File: lib/dart/lib/src/frugal/transport/f_async_transport.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/transport/f_async_transport.dart]: We can update the test to following instead of `await transport.request(null, null);` if we want to make the request params non-nullable. 

```
FContext reqCtx = FContext();
var frame = mockFrame(reqCtx, "request");
var respFrame = mockFrame(reqCtx, "response");
await transport.request(reqCtx, respFrame);
```
I have updated the same.

File: lib/dart/lib/src/frugal/transport/f_transport.dart (hunk 15 above)
[the remark this answers is not in the document]
Comment (brettkail-wk) [reply lib/dart/lib/src/frugal/transport/f_transport.dart:78]: Declaring a local variable allows non-null type promotion, which removes the need for `!` (sorry, I missed `this.`):
```
final requestSizeLimit = this.requestSizeLimit;
if (requestSizeLimit != null &&
    requestSizeLimit > 0 &&
    payload.length > requestSizeLimit) {
```
I don't know which style is preferred, so I was hoping @robbecker-wf had an opinion :-).

File: lib/dart/lib/src/frugal/transport/base_f_transport_monitor.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (robbecker-wf) [reply lib/dart/lib/src/frugal/transport/base_f_transport_monitor.dart]: I updated the mocks a bit to handle this situation. Please don't introduce a different mocking library.

File: lib/dart/lib/src/frugal/transport/f_async_transport.dart (hunk 11 above)
[the remark this answers is not in the document]
Comment (robbecker-wf) [reply lib/dart/lib/src/frugal/transport/f_async_transport.dart:73]: The logic in this file is very confusing, but can be boiled down to
It makes a request with a timeout and keeps track of a bunch of completers, whichever finishes first the result is used. The result is passed to `TMemoryTransport.fromUint8List(response);` which, if null will cause an exception. So we don't want this to be null. Whichever completer isn't used it is closed out in the finally.
All the completers and such are all typed as bytes (aka Uint8List). The code as-is looks like the best we can do.

File: lib/dart/lib/src/frugal/transport/f_transport.dart (hunk 15 above)
[the remark this answers is not in the document]
Comment (robbecker-wf) [reply lib/dart/lib/src/frugal/transport/f_transport.dart:78]: It's kinda nice to get rid of the `!` here. So that code block that Brett showed works nicely. Let's do that.

File: lib/dart/pubspec.yaml
Comment (robbecker-wf) [inline lib/dart/pubspec.yaml:1]: I couldn't take it, I had to reorder the pubspec.yaml into a normal order

Comment (alishadalal-wk) [reply lib/dart/pubspec.yaml:1]: Updated pubspec by resolving the merge conflicts.

File: lib/dart/lib/src/frugal/f_context.dart [hunk 16]
```diff
@@ -94,11 +97,11 @@ class FContext {
   }
 
   /// Correlation id for the context.
-  String get correlationId => _requestHeaders[_cidHeader];
+  String? get correlationId => _requestHeaders[_cidHeader];
 
   /// The operation id for the context.
   int get _opId {
-    var opIdStr = _requestHeaders[_opidHeader];
+    var opIdStr = _requestHeaders[_opidHeader]!;
```
Comment (danaoredson-wf) [inline lib/dart/lib/src/frugal/f_context.dart:104]: Curious if analyzer is flagging this as something that could be made `final`.

File: lib/dart/lib/src/frugal/transport/base_f_transport_monitor.dart [hunk 17]
```diff
@@ -68,18 +68,19 @@ class BaseFTransportMonitor extends FTransportMonitor {
...
-    }
-
-    return (prevWait * 2).clamp(0, _maxWait);
+  int? onReopenFailed(int? prevAttempts, int? prevWait) {
+    if (prevAttempts != null && prevWait != null) {
+      if (prevAttempts >= _maxReopenAttempts) {
+        return -1;
+      }
+      return (prevWait * 2).clamp(0, _maxWait);
+    } else
```
Comment (danaoredson-wf) [inline lib/dart/lib/src/frugal/transport/base_f_transport_monitor.dart:82]: nit: unnecessary else, since all branches of the `if` return.

File: lib/dart/lib/src/frugal/transport/f_http_transport.dart [hunk 18]
```diff
@@ -36,7 +36,7 @@ class FHttpTransport extends FTransport {
   /// No limit will be enforced if set to a non-positive value (i.e. <1).
   final int responseSizeLimit;
 
-  Map<String, String> _headers;
+  Map<String, String> _headers = {};
```
Comment (danaoredson-wf) [inline lib/dart/lib/src/frugal/transport/f_http_transport.dart:39]: Can this be `final`?

File: lib/dart/lib/src/frugal/f_context.dart
[the remark this answers is not in the document]
Comment (robbecker-wf) [reply lib/dart/lib/src/frugal/f_context.dart:1]: A follow up ticket would be best. frugal is pretty unmaintained, so it would be easy to roll in a lot of unrelated cleanup but it's best to try to keep the changes to as small of a changeset as possible.

[the remark this answers is not in the document]
Comment (danaoredson-wf) [reply lib/dart/lib/src/frugal/f_context.dart:1]: For sure.

File: lib/dart/lib/src/frugal/transport/f_http_transport.dart (hunk 18 above)
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/transport/f_http_transport.dart]: Made `Map<String, String> _headers = {};` to `final Map<String, String> _headers;` and updated the code for the same.

File: lib/dart/lib/src/frugal/transport/base_f_transport_monitor.dart (hunk 17 above)
Comment (alishadalal-wk) [reply lib/dart/lib/src/frugal/transport/base_f_transport_monitor.dart]: Removed else from here & updated the return outside if.

**Review by robbecker-wf — APPROVED** [review]

**Comment by robbecker-wf** [comment]
QA+1 CI passes. A consumer test batch https://sourcegraph.plat.workiva.net/organizations/Workiva/batch-changes/test_frugal_nullsafe was done on all 98 consumers of frugal. about half passed right off the bat. The others only failed due to
- failure to setup git ssh in the dockerfile before trying a pub get. fixing these in many repos resulted in passing.
- lockfile not up to date in the app directory
- formatting changes (due to resolving to newer analyzer)

key repos like DPC, audit, messaging-sdk all passed.

**Review by dustinlessard-wf — APPROVED** [review]

**Comment by dustinlessard-wf** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM
